File: Python/BOJ/14889.py
import sys

n = int(sys.stdin.readline())
visited = [False for _ in range(n)]
table = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
min_dif = int(1e9)


def dfs(depth, idx):
    global min_dif
    if depth == n // 2:
        power1, power2 = 0, 0
        for i in range(n):
            for j in range(n):
                if visited[i] and visited[j]:
                    power1 += table[i][j]
                elif not visited[i] and not visited[j]:
                    power2 += table[i][j]
        min_dif = min(min_dif, abs(power1 - power2))
        return

    for i in range(idx, n):
        if not visited[i]:
            visited[i] = True
            dfs(depth + 1, i + 1)
            visited[i] = False
dfs(0, 0)
print(min_dif)

File: C++/BOJ/14889.cpp
#include<iostream>
#include<cmath>

using namespace std;

int n;
int table[20]{};
int min_dif = 1e9;

void dfs(int idx, int start, int sum)
{
    if (idx == n / 2)
    {
        if (min_dif > std::abs(sum))
            min_dif = std::abs(sum);
    }
    if (start < n - 1)
    {
        dfs(idx + 1, start + 1, sum - table[start]);
        dfs(idx, start + 1, sum);
    }
}

int main()
{
    int sum = 0;
    int num;
    cin>>n;
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin>>num;
            sum += num;
            table[i] += num;
            table[j] += num;
        }
    }
    dfs(0, 0, sum);
    cout << min_dif;
}